hack/update_renovate_json_based_on_codeowners.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python
+import argparse
+import json
+import re
+from itertools import groupby
+from pathlib import Path
+from typing import Any, Iterable, Iterator, TypedDict
+
+
+class PackageRule(TypedDict):
+    groupName: str
+    matchFileNames: list[str]
+
+
+def get_renovate_packagerules(codeowners_content: str) -> Iterator[PackageRule]:
+    lines = map(str.strip, codeowners_content.splitlines())
+    rules: list[PackageRule] = []
+
+    for isempty, lines_group in groupby(lines, key=lambda line: not line):
+        if not isempty and (rule := _process_owner_group(lines_group)):
+            rules.append(rule)
+
+    rules.sort(key=lambda rule: rule["groupName"])
+
+    for groupname, rules_group in groupby(rules, key=lambda rule: rule["groupName"]):
+        merged_patterns = set()
+        for rule in rules_group:
+            merged_patterns.update(rule["matchFileNames"])
+        yield {"groupName": groupname, "matchFileNames": sorted(merged_patterns)}
+
+
+def _process_owner_group(group: Iterable[str]) -> PackageRule | None:
+    """Process a group of CODEOWNERS.
+
+    If the group has a '# renovate groupName=' directive, return a packageRules object.
+    Otherwise, return None.
+    """
+    renovate_directive_pat = re.compile(r"#\s*renovate\s+groupName=(.*)")
+
+    patterns = []
+    groupname = None
+
+    for line in group:
+        if not line.startswith("#"):
+            pattern, *_ = line.split(maxsplit=1)
+            patterns.append(pattern)
+        elif m := renovate_directive_pat.match(line):
+            groupname = m.group(1)
+
+    if groupname:
+        patterns = list(map(_codeowners_pattern_to_glob_pattern, patterns))
+        return {"groupName": groupname, "matchFileNames": patterns}
+    else:
```

Comment:
nit: the `else` is unnecessary afaik

Reply:
Yup. Also reordered to avoid making the `return None` look like the happy path :smile: